Circular header splitting

tiennou opened the pull request:
I've started to look at adding `libssh` to the party, mostly to look at how much an impedance mismatch we would get from what `libssh2` provides.

This is not really about this, but since I hit some crazy circular includes again, I attempted to separate some of the more simple, globally used things into their own header (`cert`). This also tries to make it less likely to look at low-level things — the `cred` system suffers from this, as its implementation must be public in case someone has a specific authentication system to support, but that distinction gets lost when you're down looking at `transport.h` (It also seems to miss a `get_username` member to really be "polymorphic" for those implementers).

I've done the same with the functions from `remote.h` that really are transport-related, ideally making the separation between the "main" ones from the "step-by-step" ones, but that's tentative — I might just drop that altogether.

I've tried to not break any (obvious) includes, but this slightly breaks backward compatibility depending on how much of `git2/sys/cred.h` you need to see. I'm fine with it being added to the main `cred.h` header to handle that case.

cuviper commented:
> I've started to look at adding `libssh` to the party

I'm also interested in this, per #5225. Let me know if you want to work together!

pks-t commented on include/git2/transport.h:
```diff
@@ -22,9 +24,535 @@
  */
 GIT_BEGIN_DECL
 
+/**
```
How about we move that discussion into a separate PR then to not hold off the efforts towards implementing a libssh backend? I'd love to have others have a say first before giving a definite ACK/NACK.

tiennou commented:
I've removed the change and will resubmit when ready. I've added a changelog as well, documenting the *only* thing I would expect to fail is actual custom transport implementations, or people prodding the structs directly after setting them up using the public functions, which is weird (the argument is that we're only ever asking for `git_cred` pointers via the `git_cred_acquire_cb`, and those are never initialized).

pks-t commented on include/git2/cred.h:
```diff
@@ -0,0 +1,308 @@
+ * @param cred object to check
+ * @return the credential username, or NULL if not applicable
+ */
+GIT_EXTERN(const char *) git_cred_get_username(git_cred *cred);
```
Well, I'd usually note that this is an atypical API name, as we almost never add the "get" verb. But in fact compilers will complain if there is a function and typedef with the same name:

```C
#include <stdio.h>

typedef struct git_cred_username {
    char *username;
} git_cred_username;

char *git_cred_username(const git_cred_username *c)
{
    return c->username;
}

int main(int argc, const char *argv[])
{
    git_cred_username c = { "foo" };
    printf("Username: %s\n", git_cred_username(&c));
    return 0;
}
```

Compiling this results in the following error:

```bash
$ gcc main.c
main.c:7:7: error: 'git_cred_username' redeclared as different kind of symbol
 char *git_cred_username(const git_cred_username *c)
       ^~~~~~~~~~~~~~~~~
main.c:5:3: note: previous declaration of 'git_cred_username' was here
 } git_cred_username;
   ^~~~~~~~~~~~~~~~~
main.c: In function 'main':
main.c:14:22: error: expected ';' before 'c'
     git_cred_username c = { "foo" };
                      ^~
                      ;
main.c:15:49: error: 'c' undeclared (first use in this function)
     printf("Username: %s\n", git_cred_username(&c));
                                                 ^
main.c:15:49: note: each undeclared identifier is reported only once for each function it appears in
```

So calling it `get_username` is in fact the best way forward, even though uncommon.

pks-t commented:
Thanks a lot @tiennou, keep up the good work :)

jonathanturcotte commented:
This seems to have broken the -DUSE_GSSAPI build :(

```bash
[ 20%] Building C object src/CMakeFiles/git2internal.dir/util.c.o
/home/ec2-user/libgit2/src/util.c:735:13: warning: ‘insertsort’ defined but not used [-Wunused-function]
 static void insertsort(
             ^
[ 21%] Building C object src/CMakeFiles/git2internal.dir/varint.c.o
[ 21%] Building C object src/CMakeFiles/git2internal.dir/vector.c.o
[ 21%] Building C object src/CMakeFiles/git2internal.dir/wildmatch.c.o
[ 21%] Building C object src/CMakeFiles/git2internal.dir/worktree.c.o
[ 21%] Building C object src/CMakeFiles/git2internal.dir/zstream.c.o
[ 21%] Building C object src/CMakeFiles/git2internal.dir/allocators/stdalloc.c.o
[ 22%] Building C object src/CMakeFiles/git2internal.dir/allocators/win32_crtdbg.c.o
[ 22%] Building C object src/CMakeFiles/git2internal.dir/streams/mbedtls.c.o
[ 22%] Building C object src/CMakeFiles/git2internal.dir/streams/openssl.c.o
[ 22%] Building C object src/CMakeFiles/git2internal.dir/streams/registry.c.o
[ 22%] Building C object src/CMakeFiles/git2internal.dir/streams/socket.c.o
[ 22%] Building C object src/CMakeFiles/git2internal.dir/streams/stransport.c.o
[ 23%] Building C object src/CMakeFiles/git2internal.dir/streams/tls.c.o
[ 23%] Building C object src/CMakeFiles/git2internal.dir/transports/cred_helpers.c.o
[ 23%] Building C object src/CMakeFiles/git2internal.dir/transports/git.c.o
[ 23%] Building C object src/CMakeFiles/git2internal.dir/transports/local.c.o
[ 23%] Building C object src/CMakeFiles/git2internal.dir/transports/smart.c.o
[ 24%] Building C object src/CMakeFiles/git2internal.dir/transports/smart_pkt.c.o
[ 24%] Building C object src/CMakeFiles/git2internal.dir/transports/smart_protocol.c.o
[ 24%] Building C object src/CMakeFiles/git2internal.dir/transports/auth_negotiate.c.o
In file included from /home/ec2-user/libgit2/src/common.h:31:0,
                 from /home/ec2-user/libgit2/src/transports/auth_negotiate.h:11,
                 from /home/ec2-user/libgit2/src/transports/auth_negotiate.c:8:
/home/ec2-user/libgit2/src/transports/auth_negotiate.c: In function ‘negotiate_next_token’:
/home/ec2-user/libgit2/src/transports/auth_negotiate.c:90:54: error: dereferencing pointer to incomplete type
  assert(buf && ctx && ctx->configured && cred && cred->credtype == GIT_CREDTYPE_DEFAULT);
                                                      ^
gmake[2]: *** [src/CMakeFiles/git2internal.dir/transports/auth_negotiate.c.o] Error 1
gmake[1]: *** [src/CMakeFiles/git2internal.dir/all] Error 2
gmake: *** [all] Error 2
```

I can build fine if I use commit e24b8852249716dfc032cdc38f53fbd4a5f12b6e

pks-t commented:
@jonathanturcotte: thanks for the report. This should be fixed now by #5237
